EVP_{CIPHER,MD}_CTX_ctrl(): make extra sure to return 0 or 1

levitte opened the pull request:
The previous fix (#10108) was incomplete.

Fixes #10106

levitte committed:
EVP_{CIPHER,MD}_CTX_ctrl(): make extra sure to return 0 or 1

The previous fix was incomplete.

Fixes #10106

levitte commented:
(note that I haven't really tested.  It's Sunday...)

petrovr commented:
Richard Levitte wrote:
> The previous fix (#10108) was incomplete.
>
> Fixes #10106

In pre-master EVP_CTRL_INIT returns 1. Just goto will use default value set ate beginning of function (-1).
I think that for this case code should be ret = 1; goto concluide .

Roumen

slontis commented on crypto/evp/evp_enc.c:
```diff
@@ -1090,7 +1090,7 @@ int EVP_CIPHER_CTX_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg, void *ptr)
     case EVP_CTRL_INIT: /* TODO(3.0) Purely legacy, no provider counterpart */
     default:
-        return EVP_CTRL_RET_UNSUPPORTED;
+        goto conclude;
```
Is this supposed to be ret = UNSUPPORTED here??

t8m replied:
It is already set to that value.

levitte replied:
Yup.  Look at line 1065

richsalz replied:
While you're here, please fix the inconsistency and rename the label to end or err like everywhere else in openssl.

levitte replied:
Meh :unamused: ...  but ok

levitte committed:
fixup! EVP_{CIPHER,MD}_CTX_ctrl(): make extra sure to return 0 or 1

t8m commented:
Is this still WIP?

levitte committed:
For provided ciphers, EVP_CIPHER_CTX_ctrl() with EVP_CTRL_INIT always returns 1

This control command should never be used with provided methods, but
since this is publically available, someone might still make the
mistake.  We make sure it returns 1 so as not to be overly
disruptive.

levitte commented:
@petrovr, I just added a commit that deals with EVP_CTRL_INIT as you requested.  Please confirm that this is what you thought it should do.

levitte commented:
And *now* it's time to take this out of WIP :wink:

petrovr commented:
Richard Levitte wrote:
> @petrovr, I just added a commit that deals with EVP_CTRL_INIT as you requested.  Please confirm that this is what you thought it should do.
> 
Ok.
For protocol currently it is used in crypto/evp/evp_enc.c, line 442, 
legacy code , when flag EVP_CIPH_CTRL_INIT is set and ... etc

Roumen

levitte commented:
Thanks everyone!

Merged.

6a36f209bcfbc958e30d544c3031ed002dbfe582 EVP_{CIPHER,MD}_CTX_ctrl(): make extra sure to return 0 or 1
d6d74cf4a44f08ed63d942b103198a1e3294295a For provided ciphers, EVP_CIPHER_CTX_ctrl() with EVP_CTRL_INIT always returns 1
